### DISCUSSION TITLE
Feature Validation Updates

### DISCUSSION THREAD

**Discussion by FamroFexl** [FamroFexl/Circumnavigate#4] [Show and tell]
# What are the recent developments with features?

**Comment by FamroFexl** [comment]
Just validated that previously client-handled functions can now be handled by the server, enabling vanilla client support for world wrapping! Amazing!

**Reply by FamroFexl** [reply]
This will increase the computational load on the server, but I think vanilla client support is worth it! Maybe requiring clients who use mod loaders to have Circumnavigate installed will help unburden the server, as they can handle their own calculations. 

I still haven't profiled the additional overhead all of the mixins and coordinate transformations will incur.

**Comment by FamroFexl** [comment]
Validated that 4D noise generation will work for producing seamless edges in regularly-wrapped world. It is yet to be known if the same applies for chunk-shifted worlds.

The image below is a 4D-generated image tiled into a 4x4 with green dividers. Notice the seamless intersection between the blob sections.
![Untitled](https://github.com/user-attachments/assets/55defc34-6df4-4332-a086-97a51a4bb45d)

**Reply by FamroFexl** [reply]
From what I understand, 4D perlin noise is significantly more computationally expensive for world generation than vanilla's traditional 2D perlin noise. This computational load will also rise due to the new 3D cave features, which will require a 5D perlin noise implementation.

**Reply by FamroFexl** [reply]
> From what I understand, 4D perlin noise is significantly more computationally expensive for world generation than vanilla's traditional 2D perlin noise. This computational load will also rise due to the new 3D cave features, which will require a 5D perlin noise implementation.

Noise generation algorithms provided by https://github.com/KdotJPG/OpenSimplex2. 5D noise generation might have to be commissioned from @KdotJPG or other mathematicians.

**Reply by FamroFexl** [reply]
> From what I understand, 4D perlin noise is significantly more computationally expensive for world generation than vanilla's traditional 2D perlin noise. This computational load will also rise due to the new 3D cave features, which will require a 5D perlin noise implementation.

This might just be hearsay. It might just be that Y values will wrap as they have in the rest of the mod.

**Comment by FamroFexl** [comment]
Using an [Authgraph](http://www.authagraph.com/top/?lang=en) world projection (which is the most accurate 2d tileable world map projection I know of), it is possible to create fully trans-navigable earth minecraft worlds.
![kunimunes-authagraph-4096p-incl-polar-icecaps](https://github.com/user-attachments/assets/7dc1c82c-fe83-4d14-a2b5-e12c2d47f08d)

**Reply by FamroFexl** [reply]
Now as you can see, it isn't as straightforward a mapping process as first believed. For instance, chunks will have to be flipped and switched in order to accommodate this system.
![AG_EN_04-638x479](https://github.com/user-attachments/assets/66a3f446-fd6e-45e6-87b1-d5e9c2df88fb)

**Reply by FamroFexl** [reply]
This system probably won't work since the same entity will be next to themselves on the edge of the northern hemisphere. I am literally beside myself.

**Reply by Skidamek** [reply]
Yeah, maybe since it's Minecraft our blocky game, maybe world in in shape of cube would be cool instead of trying to make it similar to earth. But that would require probably a lot more calculations in wrapping/unwrapping logic. Because currently our world is just one face of cube connected infinitely, the cube shape would require 5 more of those faces.

**Reply by FamroFexl** [reply]
Well, a cube earth is impossible to implement since the corners connect at 3 faces, not 4, making it impossible to flatten out.

**Reply by Skidamek** [reply]
Oh you're right

**Comment by FamroFexl** [comment]
### Circumnavigate Modded-Client Exclusive Features.

- Duplicate Chunk Loading for increased view distances
 
  This is technically possible for a vanilla client, but it would incur so many complications on behalf of the server that it is unreasonable for the server to support it.
  The modded vanilla client will have to Duplicate chunks, entities, explosions, particles, and all other events to maintain consistency.

- 2D View Distances

  In worlds where the axis limits don't make up a square, the client can load more chunks in one direction than in another direction. However, vanilla client's only support a square view distance, so they can only be as big as the minimum axis. In the modded client, it can be given separate values for the maximum view distances avaliable on different axis.

**Comment by FamroFexl** [comment]
Wrapped Structure and Feature generation appears to be impossible with Mixin changes alone. This will delay and possibly eliminate the chance of wrapped structure and feature generation occurring. Before I can describe why this is, an explanation of structure generation is listed below, based on all the info I obtained.

1. During the `STRUCTURE_STARTS` chunk generation step, Minecraft decides which chunks will become the starting point of a structure, defined by a `StructureStart`. A `StructureStart` determines what structure will be generated, the pieces of that structure, the `BoundingBox` which encompasses the entire structure, and the `ChunkPos` it starts from. For instance, a village is based around a single chunk which holds a `StructureStart`.

2. Later, in the `STRUCTURE_REFERENCES` chunk generation step, other chunks are identified as being part of a structure in the area based on their intersection with the `BoundingBox` of a `StructureStart` as expressed in the previous step. They receive a pointer to a `StructureStart` in this process.

3. Lastly, during the `FEATURES` generation step, the structures are actually generated. Chunks use the pointers they obtained from the `STRUCTURE_REFERENCES` step to generate portions of a structure, depending on the parameters obtained from their linked `StructureStart`.

Importantly, throughout this entire process, Minecraft groups sections of chunks together in a `StaticCache2D` cache, operated by a `WorldGenRegion` (distinct from region files). When generating, `StructureStarts` are only capable of operating on chunks within their chunk cache. This means that in order for a single structure to generate on both sides of the world, the chunk cache needs to contain generating chunks that aren't next to each other, or even nearby. In order to do that, the entire generation system would have to be overhauled in order to accommodate this, and that is a task much bigger and more complicated than simply switching out values in a few mixins.
